### ISSUE TITLE
All handle parameters being abstract SafeHandle punishes the use of GetCursorInfo and GetIconInfo

### ISSUE THREAD

**Issue by jnm2** [microsoft/CsWin32#125]
No semantically correct SafeHandle type is provided by CsWin32 now that all handle parameters are `SafeHandle`:

```cs
var cursorInfo = new CURSORINFO { cbSize = Marshal.SizeOf(typeof(CURSORINFO)) };

if (!PInvoke.GetCursorInfo(ref cursorInfo))
    throw new Win32Exception();

//                           ↓ Options: abuse an existing SafeHandle type or declare one myself
if (!PInvoke.GetIconInfo(new WhatDoIWriteHere(cursorInfo.hCursor, ownsHandle: false), out var iconInfo))
    throw new Win32Exception();
```

Prior to everything becoming abstract SafeHandle, this code worked which isn't beautiful but is nicer than the two new options:

```cs
if (!PInvoke.GetIconInfo(new(cursorInfo.hCursor, ownsHandle: false), out var iconInfo))
    throw new Win32Exception();
```

I'm also not a fan of throwing away the parameter type information in this way. `GetObject` has parameter `SafeHandle h` now, for instance. A parameter like that doesn't help you figure out what kind of handle to pass or how to get it.

**Comment by jnm2** [comment]
`new UnownedHandle(cursorHandle)` is not so bad, syntax-wise, though I'm still uneasy about the loss of signature information.

**Comment by AArnott** [comment]
Yes, the type-specificity is valuable, I agree. And we may bring it back, when we get the majority of the APIs "correct" and can circle back to bring back the niceties. We can leave this issue open to track considering bringing those back.
We may offer non-SafeHandle overloads as well, since that sounds like something that will help in some cases.

**Comment by AraHaan** [comment]
I would also like overloads generated to generate overloads for ``HandleRef`` as well too.

**Comment by AArnott** [comment]
@arahaan: `HandleRef` has been [deprecated since .NET 2.0](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.handleref?view=net-5.0). We have to keep the number of overloads we generate to a very small number to maintain an overall good experience, so they have to appeal to a broad audience. I think such a very old type won't meet the bar.
But if you'd like to explain why this is important to you, I'd like to hear it.

**Comment by AraHaan** [comment]
``HandleRef`` is not only used by my application (for referencing HWND handles to winforms controls to pass them into p/invokes).

They are also used in the runtime itself (if I remember right they are used a lot in the winforms repository, maybe also the wpf repository in the dotnet org.

Possibly used inside of ``System.Drawing``, ``System.Drawing.Common`` as well too. One way to find out is to search ``HandleRef`` in github under the ``dotnet`` org scope to see where it all is used.

**Comment by jnm2** [comment]
HandleRef is useful for when you don't own the handle, but you did create the managed object that owns the handle and you need to keep that object from being collected and freeing the handle concurrently during the native call.

Since HandleRef is a struct, it won't ever introduce an overload resolution ambiguity when passing the argument `null`.

**Comment by AArnott** [comment]
Alright. I guess it may be worth considering `HandleRef` overloads. But please file an issue to track that because it's decidedly off-topic here.

**Comment by AraHaan** [comment]
I think generating ``SafeHandle`` overloads should be opt-in by default but have the user opt-in for them if they specifically ask for said overload to be generated within the configuration for specific members only (opt-in per p/invoke function name) within the text file that controls the source generator.

**Comment by AArnott** [comment]
@AraHaan I tend to disagree since most APIs that produce handles expect the caller to close those handles, and .NET has a very strong precedent for using `SafeHandle` here, and most users want that. Our defaults should match the expectations of most folks.
We might expose an *option* to suppress or alter friendly overload generation though, which may give you what you want.

**Comment by AArnott** [comment]
One issue with declaring input parameters with specific `SafeHandle`-derived types is the existence of multiple derived types that should all be valid inputs. .NET itself defines multiple `SafeHandle`-derived types that close with `CloseHandle`. By declaring parameters as taking a `SafeHandle`, all these are acceptable. If we forced them to take `SafeFileHandle` (which is just one such type that closes with `CloseHandle`), then all the other handle types wouldn't work as arguments, even though they were legitimately acquired from other APIs.

**Comment by jevansaks** [comment]
To solve this problem and #1393, I propose two changes:
1. All generated SafeHandles have an implicit cast operator to their underlying "HANDLE" type (e.g. HMODULE for FreeLibrarySafeHandle, BSTR for SysFreeStringSafeHandle, etc).
2. All friendly overloads with [In] parameters take the underlying HANDLE type as [In] parameter so they can accept either the SafeHandle wrapper or the raw HANDLE. 

Thoughts @AArnott ?

**Comment by jevansaks** [comment]
After talking with @AArnott offline, I realized that SafeHandle auto-casting to HANDLE can't work because of the pattern for DangerousGetAddRef/Release that you need to do in the marshalling code. But maybe should introduce a holder struct that can take either HANDLE or SafeHandle to make it easier to pass HANDLE in (or BSTR, PCWSTR, PVOID, etc as the case may be).
